raiden-ts/src/transfers/epics/withdraw.ts

```diff
@@ -405,3 +406,155 @@ export function withdrawSendExpireMessageEpic(
+          ),
+        );
+      }).pipe(
+        catchError((error) => {
+          log.warn('Could not reply to CoopSettle request, ignoring', { action, error });
+          return EMPTY;
+        }),
+      ),
+    ),
+  );
+}
+
+/**
+ * When both valid [[WithdrawConfirmation]] for a [[WithdrawRequest]].coop_settle=true from us,
+ * send a channelSettle.request
+ *
+ * @param action$ - Observable of withdrawMessage.success actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.config$ - Config observable
+ * @param deps.log - Logger instance
+ * @returns Observable of channelSettle.request actions
+ */
+export function coopSettleEpic(
+  action$: Observable<RaidenAction>,
+  state$: Observable<RaidenState>,
+  { config$, log }: RaidenEpicDeps,
+): Observable<channelSettle.request | withdraw.failure> {
+  return action$.pipe(
+    filter(withdrawMessage.success.is),
+    withLatestFrom(state$, config$),
+    map(([action, state, { revealTimeout }]) => {
+      // don't act if request expires too soon
+      if (action.meta.expiration < state.blockNumber + revealTimeout) return;
+
+      const channel = state.channels[channelKey(action.meta)];
+      if (channel?.state !== ChannelState.open) return;
+
+      const { ownCapacity, partnerCapacity, ownTotalWithdrawable, partnerTotalWithdrawable } =
+        channelAmounts(channel);
+      if (
+        channel.own.locks.length ||
+        channel.partner.locks.length ||
+        !ownCapacity.isZero() || // when both capacities are zero, both sides should be ready
+        !partnerCapacity.isZero()
+      )
+        return;
+
+      const ownReq = channel.own.pendingWithdraws.find(
+        (msg): msg is Signed<WithdrawRequest> =>
+          msg.type === MessageType.WITHDRAW_REQUEST &&
+          msg.expiration.gte(state.blockNumber + revealTimeout) &&
+          msg.total_withdraw.eq(ownTotalWithdrawable) &&
+          !!msg.coop_settle, // only requests where coop_settle is true
+      );
+      if (!ownReq) return; // not our request
+
+      const ownConfirmation = channel.own.pendingWithdraws.find(
+        matchWithdraw(MessageType.WITHDRAW_CONFIRMATION, ownReq),
+      );
+      const partnerReq = channel.partner.pendingWithdraws.find(
+        (msg): msg is Signed<WithdrawRequest> =>
+          msg.type === MessageType.WITHDRAW_REQUEST &&
+          msg.expiration.gte(state.blockNumber + revealTimeout) &&
+          msg.total_withdraw.eq(partnerTotalWithdrawable),
+      );
+      if (!partnerReq) return; // shouldn't happen
+      const partnerConfirmation = channel.partner.pendingWithdraws.find(
+        matchWithdraw(MessageType.WITHDRAW_CONFIRMATION, partnerReq),
+      );
+      if (!ownConfirmation || !partnerConfirmation) {
+        log.info('no matching WithdrawConfirmations found', {
+          ownConfirmation,
+          partnerConfirmation,
+        });
+        return;
+      }
+
+      return channelSettle.request(
```

Comment:
Why did you decide to include this into the `channelSettle` action? Wouldn't it be cleaner to separate these two paths?

Reply:
Mostly for semantic organisation: this is settling afterall. Also, there're some common shared logic there, although mostly some also shared with several other contract-interacting epics, but I thought someone not familiar with the codebase would look in the `settle.ts` epic module to find how `cooperativeSettle` is performed in the SC, and therefore tried to put it in the first place someone would look for it. Being made through withdraw messages is just a transport/protocol implementation detail, and therefore only the protocol part of it is here in the `withdraw.ts` module.